### PR TITLE
Make privateKey optional in signTransaction

### PR DESCRIPTION
**Pull request by jimni1222**
## Proposed changes

This PR introduces the logic to change signTransaction's privateKey to optional.

As a result of this change, if signTransaction function receives an array of keys in the privateKey parameter, the signTransaction function will sign all the keys in the array.

The return value also adds a field called signature to return the signed result.

At this time, legacy is [v, r, s], and transaction type except legacy has two-dimensional array like [[v, r, s], [v, r, s]].

The v, r, s of the existing result represents the signature of the 0th index of signature.

The main changes are as follows:
1. caver.klay.accounts.wallet.getAccount -> get account from wallet method
2. parse arguments in signTransaction for handling optional parameters(privateKey, callback)
3. sign to transaction with array of private key

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/caver-js/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/caver-js)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: packages/caver-klay/caver-klay-accounts/src/index.js [hunk 1]
```diff
@@ -191,50 +191,85 @@ Accounts.prototype.getLegacyAccount = function getLegacyAccount(key) {
...
+      if (callback) callback(e)
+      return Promise.reject(e)
+    }
 
-    if (!tx) {
-      error = new Error('No transaction object given!')
-      callback(error)
-      return Promise.reject(error)
+    if (arguments.length === 0 || arguments.length > 3) {
+      return handleError('Inavlid parameter: The number of parameter is invalid.')
```
Comment (kjhman21) [inline packages/caver-klay/caver-klay-accounts/src/index.js]:
```suggestion
      return handleError('Invalid parameter: The number of parameters is invalid.')
```

File: packages/caver-klay/caver-klay-accounts/src/index.js [hunk 2]
```diff
@@ -191,50 +191,85 @@ Accounts.prototype.getLegacyAccount = function getLegacyAccount(key) {
...
+    let tx = arguments[0], privateKey
 
-    try {
-      parsed = utils.parsePrivateKey(privateKey)
-      privateKey = parsed.privateKey
-    } catch(e) {
-      callback(e)
-      return Promise.reject(e)
+    if (!tx || !_.isObject(tx)) {
+      return handleError('Invalid paramter: The transaction must be defined as an object')
```
Comment (kjhman21) [inline packages/caver-klay/caver-klay-accounts/src/index.js]:
```suggestion
      return handleError('Invalid parameter: The transaction must be defined as an object')
```

File: packages/caver-klay/caver-klay-accounts/src/index.js [hunk 3]
```diff
@@ -191,50 +191,85 @@ Accounts.prototype.getLegacyAccount = function getLegacyAccount(key) {
...
+
+    if (arguments.length === 2) {
+      if (_.isFunction(arguments[1])) {
+        callback = arguments[1]
+      } else {
+        privateKey = arguments[1]
+      }
+    } else if (arguments.length === 3) {
+      if (typeof arguments[1] !== 'string' && !_.isArray(arguments[1])){
+        return handleError('Invalid paramter: The parameter for the private key is invalid')
```
Comment (kjhman21) [inline packages/caver-klay/caver-klay-accounts/src/index.js]:
```suggestion
        return handleError('Invalid parameter: The parameter for the private key is invalid')
```

File: packages/caver-klay/caver-klay-accounts/src/index.js [hunk 4]
```diff
@@ -191,50 +191,85 @@ Accounts.prototype.getLegacyAccount = function getLegacyAccount(key) {
...
+        privateKey = account.privateKey
+      } catch(e) {
+        return handleError(e)
       }
-      if (error) {
-        callback(error);
-        return Promise.reject(error);
+    }
+
+    let privateKeys = _.isArray(privateKey) ? privateKey : [privateKey]
```
Comment (kjhman21) [inline packages/caver-klay/caver-klay-accounts/src/index.js]: _.isArray(privateKey)?

Is this right? I think privateKey is always an array like you checked before at line 225:
```
_.isArray(privateKey)
```

Comment (jimni1222) [reply packages/caver-klay/caver-klay-accounts/src/index.js]: In line 225, if privateKey is not string or array, return error.
So, privateKey can be a string type. :)

File: packages/caver-klay/caver-klay-accounts/src/index.js [hunk 5]
```diff
@@ -191,50 +191,85 @@ Accounts.prototype.getLegacyAccount = function getLegacyAccount(key) {
...
+        privateKeys[i] = parsed.privateKey
+        privateKeys[i] = utils.addHexPrefix(privateKeys[i])
+  
+        if (!utils.isValidPrivateKey(privateKeys[i])) return handleError('Invalid private key')
       }
+    } catch(e) {
+      return handleError(e)
+    }
 
+    // Attempting to sign with decoupled account into a legacy type transaction should be rejected.
```
Comment (kjhman21) [inline packages/caver-klay/caver-klay-accounts/src/index.js]:
```suggestion
    // Attempting to sign with a decoupled account into a legacy type transaction should be rejected.
```

File: packages/caver-klay/caver-klay-accounts/src/index.js [hunk 6]
```diff
@@ -245,16 +280,22 @@ Accounts.prototype.signTransaction = function signTransaction(tx, privateKey, ca
 
         const messageHash = Hash.keccak256(rlpEncoded)
 
-        const signature = Account.makeSigner(Nat.toNumber(transaction.chainId || "0x1") * 2 + 35)(messageHash, privateKey)
-        const [v, r, s] = Account.decodeSignature(signature).map(sig => utils.makeEven(utils.trimLeadingZero(sig)))
+        let signatrues = []
```
Comment (kjhman21) [inline packages/caver-klay/caver-klay-accounts/src/index.js]:
```suggestion
        let signatures = []
```

Comment (jimni1222) [reply packages/caver-klay/caver-klay-accounts/src/index.js]: oh.... there are many mistypo ;-;.. i will edit follow logic with this.

File: packages/caver-klay/caver-klay-accounts/src/index.js [hunk 7]
```diff
@@ -245,16 +280,22 @@ Accounts.prototype.signTransaction = function signTransaction(tx, privateKey, ca
...
 
-        const rawTransaction = makeRawTransaction(rlpEncoded, [v, r, s], transaction)
+        for(const privateKey of privateKeys) {
+          const signature = Account.makeSigner(Nat.toNumber(transaction.chainId || "0x1") * 2 + 35)(messageHash, privateKey)
+          const [v, r, s] = Account.decodeSignature(signature).map(sig => utils.makeEven(utils.trimLeadingZero(sig)))
+          signatrues.push([v, r, s])
```
Comment (kjhman21) [inline packages/caver-klay/caver-klay-accounts/src/index.js]:
```suggestion
          signatures.push([v, r, s])
```

Comment (jimni1222) [reply packages/caver-klay/caver-klay-accounts/src/index.js]: I edited this mistypo :) Can you refresh PR page?

File: packages/caver-klay/caver-klay-accounts/src/index.js (hunk 4 above)
Comment (kjhman21) [reply packages/caver-klay/caver-klay-accounts/src/index.js]: Also, privateKey can be an array, too, based on the condition:

```js
 if (typeof arguments[1] !== 'string' && !_.isArray(arguments[1])){
```

Comment (jimni1222) [reply packages/caver-klay/caver-klay-accounts/src/index.js]: Yes sure, privateKey can be an array too ! 
So this line is for converting privatekey to Array.
Check privateKey is an Array or not, and if not, make privateKey to Array type.

Comment (kjhman21) [reply packages/caver-klay/caver-klay-accounts/src/index.js]: Got it.

**Comment by jimni1222** [comment]
This PR has to be rebased on dev branch after https://github.com/klaytn/caver-js/pull/110 is merged.

File: packages/caver-klay/caver-klay-accounts/src/makeRawTransaction.js [hunk 8]
```diff
@@ -144,7 +144,7 @@ function makeRawTransaction(rlpEncoded, sig, transaction) {
     }
     case 'LEGACY':
     default:
-      rawTx = decodedValues.slice(0, 6).concat(sig)
+      rawTx = decodedValues.slice(0, 6).concat(sig[0])
```
Comment (kjhman21) [inline packages/caver-klay/caver-klay-accounts/src/makeRawTransaction.js]: Just wondering, you returned first signatures when it is legacy transaction
at line 298 
https://github.com/klaytn/caver-js/pull/107/files#diff-e6b221946c9583fd0a407bd90c74ef39R298

In that case, you need to use sig, not sig[0].

Comment (jimni1222) [reply packages/caver-klay/caver-klay-accounts/src/makeRawTransaction.js]: This makeRawTransaction function is called in signTransaction when make rawTrasnaction.
In signTransaction, just send signatures array to this function.
So in makeRawTransaction function, if transaction is legacy, just use 0th signature of signatures.

**Pull request merged**